fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderHelper.java

```diff
@@ -0,0 +1,168 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.rule.join;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.starrocks.analysis.JoinOperator;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.Projection;
+import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.Map;
+
+public class JoinReorderHelper {
+
+    private JoinReorderHelper() {
+        //not called
+    }
+
+    /*
+     * split bottomJoin projection.
+     * idx 0 stores the cols should be split into the projection of newBottomJoin.
+     * idx 1 stores the cols should be split into the projection of the other child of newTopJoin.
+     * This helps to keep const cols stay in same level.
+     */
+    public static List<Map<ColumnRefOperator, ScalarOperator>> splitProjection(
+            Projection projection, ColumnRefSet colSetBottom) {
+        Map<ColumnRefOperator, ScalarOperator> top = Maps.newHashMap();
+        Map<ColumnRefOperator, ScalarOperator> bottom = Maps.newHashMap();
+        if (projection != null) {
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.getColumnRefMap().entrySet()) {
+                // To handle mappings of expressions in projection, special processing is needed like
+                // ColumnRefOperator -> ColumnRefOperator mappings with name ("1: expr" -> "2: column_name", or "1: sum" -> "2: sum"),
+                // it needs to be handled like expression mapping.
+                String keyIdentifier = entry.getKey().toString();
+                String valueIdentifier = entry.getValue().toString();
+                if (StringUtils.equals(keyIdentifier, valueIdentifier)) {
```

Comment:
why not use `equals` but the `toString().equals()` ?

Reply:
I think you are right. In my mind, this original logic is mainly used to extract projection not normal colRef-> colRef but some colRef -> expr. I will change and test it.